crates/proof-of-sql/src/proof_primitive/dynamic_dory_and_hyrax_common_utils/standard_basis_helper.rs

```diff
@@ -1,18 +1,56 @@
-//! This module provides the `build_standard_basis_vecs` method, which is used in converting a point to a
-//! vector used in a Vector-Matrix-Vector product in the dynamic dory scheme.
+use crate::base::scalar::Scalar;
+use alloc::{vec, vec::Vec};
+use core::ops::Mul;
+use itertools::Itertools;
 
-use super::F;
-use ark_ff::{AdditiveGroup, Field};
+/// Compute the hi and lo vectors (or L and R) that are derived from `point`.
+/// L and R are the vectors such that LMR is exactly the evaluation of `a` at the point `point`.
+/// # Panics
+/// This function requires that `point` has length at least as big as the number of rows in `M` that is created by `a`.
+pub(crate) fn compute_dynamic_vecs<
+    S: Scalar,
+    I: Into<S> + From<S> + Copy + core::ops::MulAssign + core::ops::Mul<Output = I>,
+>(
+    point: &[I],
+) -> (Vec<I>, Vec<I>) {
+    let point = point.iter().map(|i| (*i).into());
+    let nu = point.len() / 2 + 1;
+    let mut lo_vec = vec![I::from(S::ZERO); 1 << nu];
+    let mut hi_vec = vec![I::from(S::ZERO); 1 << nu];
+    lo_vec[0] = point
+        .clone()
+        .take(nu)
+        .map(|b| S::ONE - b)
+        .product::<S>()
+        .into();
+    hi_vec[0] = point
+        .clone()
+        .skip(nu)
+        .map(|b| S::ONE - b)
+        .product::<S>()
+        .into();
+    let standard_basis_point: Vec<I> = point
+        .map(|b| I::from((S::ONE - b).inv().expect("Values in point cannot be 1.") - S::ONE))
+        .collect_vec();
+    compute_dynamic_standard_basis_vecs::<S, _>(&standard_basis_point, &mut lo_vec, &mut hi_vec);
+    (lo_vec, hi_vec)
+}
 
-#[allow(dead_code)]
-/// This method produces evaluation vectors from a point. This is a helper method for generating a Vector-Matrix-Vector product in the dynamic dory scheme.
+/// This method produces evaluation vectors from a point. This is a helper method for generating a Vector-Matrix-Vector product in the dynamic dory and hyrax schemes.
 ///
 /// The ith element of the `lo_vec` is essentially the ith monomial basis element (lexicographically).
 /// The ith element of the `hi_vec` is essentially the jth monomial basis element where `j = row_start_index(i)`.
 ///
 /// NOTE: the `lo_vec` and `hi_vec` are scaled by `lo_vec[0]` and `hi_vec[0]` respectively.
 /// NOTE: `lo_vec` and `hi_vec` should otherwise consist entirely of zeros in order to ensure correct output.
-pub(super) fn compute_dynamic_standard_basis_vecs(point: &[F], lo_vec: &mut [F], hi_vec: &mut [F]) {
+fn compute_dynamic_standard_basis_vecs<
+    S: Scalar,
+    I: Into<S> + From<S> + Copy + core::ops::MulAssign + core::ops::Mul<Output = I>,
+>(
+    point: &[I],
+    lo_vec: &mut [I],
+    hi_vec: &mut [I],
```

Comment:
I'm not sure why I did two mutable slices rather than returning two `Vec`s. Perhaps because `(Vec<_>, Vec<_>)` is ugly. Perhaps some other reason. But, mutating slices is also not great. I wonder if there is a better solution here.